src/resolution/resolver.ts

```diff
@@ -49,89 +51,182 @@ const _resolveRequest = (requestScope: interfaces.RequestScope) =>
 
-        if (isSingleton && binding.activated) {
-            return binding.cache;
-        }
+        const exists = findExistingInScope(binding, requestScope);
 
-        if (
-            isRequestSingleton &&
-            requestScope !== null &&
-            requestScope.has(binding.id)
-        ) {
-            return requestScope.get(binding.id);
+        if (exists) {
+            return exists;
         }
 
-        if (binding.type === BindingTypeEnum.ConstantValue) {
-            result = binding.cache;
-        } else if (binding.type === BindingTypeEnum.Function) {
-            result = binding.cache;
-        } else if (binding.type === BindingTypeEnum.Constructor) {
-            result = binding.implementationType;
-        } else if (binding.type === BindingTypeEnum.DynamicValue && binding.dynamicValue !== null) {
-            result = invokeFactory(
-                "toDynamicValue",
-                binding.serviceIdentifier,
-                () => (binding.dynamicValue as (context: interfaces.Context) => any)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Factory && binding.factory !== null) {
-            result = invokeFactory(
-                "toFactory",
-                binding.serviceIdentifier,
-                () => (binding.factory as interfaces.FactoryCreator<any>)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Provider && binding.provider !== null) {
-            result = invokeFactory(
-                "toProvider",
-                binding.serviceIdentifier,
-                () => (binding.provider as interfaces.Provider<any>)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Instance && binding.implementationType !== null) {
-            result = resolveInstance(
-                binding.implementationType,
-                childRequests,
-                _resolveRequest(requestScope)
-            );
-        } else {
-            // The user probably created a binding but didn't finish it
-            // e.g. container.bind<T>("Something"); missing BindingToSyntax
-            const serviceIdentifier = getServiceIdentifierAsString(request.serviceIdentifier);
-            throw new Error(`${ERROR_MSGS.INVALID_BINDING_TYPE} ${serviceIdentifier}`);
-        }
+        let result = convertBindingToInstance(requestScope, request);
 
         // use activation handler if available
         if (typeof binding.onActivation === "function") {
             result = binding.onActivation(request.parentContext, result);
         }
 
-        // store in cache if scope is singleton
-        if (isSingleton) {
-            binding.cache = result;
-            binding.activated = true;
+        afterResult(binding, result, requestScope);
+
+        return result;
+    }
+
+};
+
+function convertBindingToInstance(requestScope: interfaces.RequestScope, request: interfaces.Request) {
```

Comment:
Use arrow functions rather than `function`

Reply:
reasoning? the prior functions in this file follow this format